File: script-files/measures/outputsContinuous/frechet.js
jStat = require("../node_modules/jstat");

dists = {
    frechet: {
        measurements: {
            nonCentralMoments: function (k, alpha, loc, scale) {
                return jStat.gammafn(1 - k / alpha)
            },
            centralMoments: function (k, alpha, loc, scale) {
                const miu1 = this.nonCentralMoments(1, alpha, loc, scale);
                const miu2 = this.nonCentralMoments(2, alpha, loc, scale);
                const miu3 = this.nonCentralMoments(3, alpha, loc, scale);
                const miu4 = this.nonCentralMoments(4, alpha, loc, scale);

                let result;
                switch (k) {
                    case 1: result = 0; break;
                    case 2: result = miu2 - miu1 ** 2; break;
                    case 3: result = miu3 - 3 * miu1 * miu2 + 2 * miu1 ** 3; break;
                    case 4: result = miu4 - 4 * miu1 * miu3 + 6 * (miu1 ** 2) * miu2 - 3 * (miu1 ** 4); break;
                };
                return result
            },
            stats: {
                mean: function (alpha, loc, scale) {
                    const miu1 = dists.frechet.measurements.nonCentralMoments(1, alpha, loc, scale);
                    return loc + scale * miu1;
                },
                variance: function (alpha, loc, scale) {
                    const miu1 = dists.frechet.measurements.nonCentralMoments(1, alpha, loc, scale);
                    const miu2 = dists.frechet.measurements.nonCentralMoments(2, alpha, loc, scale);
                    return (scale ** 2) * (miu2 - miu1 ** 2);
                },
                standardDeviation: function (alpha, loc, scale) {
                    return Math.sqrt(this.variance(alpha, loc, scale));
                },
                skewness: function (alpha, loc, scale) {
                    const central_miu3 = dists.frechet.measurements.centralMoments(3, alpha, loc, scale);
                    const miu1 = dists.frechet.measurements.nonCentralMoments(1, alpha, loc, scale);
                    const miu2 = dists.frechet.measurements.nonCentralMoments(2, alpha, loc, scale);
                    const std = Math.sqrt(miu2 - miu1 ** 2);
                    return central_miu3 / (std ** 3);
                },
                kurtosis: function (alpha, loc, scale) {
                    const central_miu4 = dists.frechet.measurements.centralMoments(4, alpha, loc, scale);
                    const miu1 = dists.frechet.measurements.nonCentralMoments(1, alpha, loc, scale);
                    const miu2 = dists.frechet.measurements.nonCentralMoments(2, alpha, loc, scale);
                    const std = Math.sqrt(miu2 - miu1 ** 2);
                    return central_miu4 / (std ** 4);
                },
                median: function (alpha, loc, scale) {
                    return dists.frechet.measurements.ppf(0.5, alpha, loc, scale);
                },
                mode: function (alpha, loc, scale) {
                    return loc + scale * (alpha / (alpha + 1)) ** (1 / alpha);
                },
            },
        }
    }
}
console.log(dists.frechet.measurements.stats.mean(5, 10, 20))
console.log(dists.frechet.measurements.stats.variance(5, 10, 20))
console.log(dists.frechet.measurements.stats.standardDeviation(5, 10, 20))
console.log(dists.frechet.measurements.stats.skewness(5, 10, 20))
console.log(dists.frechet.measurements.stats.kurtosis(5, 10, 20))
// console.log(dists.frechet.measurements.stats.median(5, 10, 20))
console.log(dists.frechet.measurements.stats.mode(5, 10, 20))

// mean_value: 33.28459427450608
// variance_value: 53.50456899676574
// standard_deviation_value: 7.314681742684759
// skewness_value: 3.5350716046214425
// kurtosis_value: 48.091512125816415
// median_value: 31.5211217027801
// mode_value: 29.283850080052545

File: phitter/continuous/continuous_distributions/frechet.py
import numpy
import scipy.optimize
import scipy.stats
import scipy.special


class Frechet:
    """
    Fréchet Distribution
    Also known as inverse Weibull distribution (Scipy name)
    https://phitter.io/distributions/continuous/frechet
    """

    def __init__(
        self,
        parameters: dict[str, int | float] = None,
        continuous_measures=None,
        init_parameters_examples=False,
    ):
        """
        Initializes the Frechet Distribution by either providing a Continuous Measures instance [ContinuousMeasures] or a dictionary with the distribution's parameters.
        Parameters Frechet Distribution: {"alpha": *, "loc": *, "scale": *}
        https://phitter.io/distributions/continuous/frechet
        """
        if continuous_measures is None and parameters is None and init_parameters_examples == False:
            raise ValueError(
                "You must initialize the distribution by providing one of the following: distribution parameters, a Continuous Measures [ContinuousMeasures] instance, or by setting init_parameters_examples to True."
            )
        if continuous_measures != None:
            self.parameters = self.get_parameters(continuous_measures=continuous_measures)
        if parameters != None:
            self.parameters = parameters
        if init_parameters_examples:
            self.parameters = self.parameters_example

        self.alpha = self.parameters["alpha"]
        self.loc = self.parameters["loc"]
        self.scale = self.parameters["scale"]

    @property
    def name(self):
        return "frechet"

    @property
    def parameters_example(self) -> dict[str, int | float]:
        return {"alpha": 5, "loc": 9, "scale": 21}

    def cdf(self, x: float | numpy.ndarray) -> float | numpy.ndarray:
        """
        Cumulative distribution function
        """
        result = scipy.stats.invweibull.cdf(x, self.alpha, loc=self.loc, scale=self.scale)
        return result

    def pdf(self, x: float | numpy.ndarray) -> float | numpy.ndarray:
        """
        Probability density function
        """
        # result = (self.alpha / self.scale) * (((x - self.loc) / self.scale) ** (-1 - self.alpha)) * numpy.exp(-(((x - self.loc) / self.scale) ** (-self.alpha)))
        result = scipy.stats.invweibull.pdf(x, self.alpha, loc=self.loc, scale=self.scale)
        return result

    def ppf(self, u: float | numpy.ndarray) -> float | numpy.ndarray:
        """
        Percent point function. Inverse of Cumulative distribution function. If CDF[x] = u => PPF[u] = x
        """
        result = self.loc + self.scale * (-numpy.log(u)) ** (-1 / self.alpha)
        return result

    def sample(self, n: int, seed: int | None = None) -> numpy.ndarray:
        """
        Sample of n elements of ditribution
        """
        if seed:
            numpy.random.seed(seed)
        return self.ppf(numpy.random.rand(n))

    def non_central_moments(self, k: int) -> float | None:
        """
        Parametric no central moments. µ[k] = E[Xᵏ] = ∫xᵏ∙f(x) dx
        """
        return scipy.special.gamma(1 - k / self.alpha)

    def central_moments(self, k: int) -> float | None:
        """
        Parametric central moments. µ'[k] = E[(X - E[X])ᵏ] = ∫(x-µ[k])ᵏ∙f(x) dx
        """
        µ1 = self.non_central_moments(1)
        µ2 = self.non_central_moments(2)
        µ3 = self.non_central_moments(3)
        µ4 = self.non_central_moments(4)

        if k == 1:
            return 0
        if k == 2:
            return µ2 - µ1**2
        if k == 3:
            return µ3 - 3 * µ1 * µ2 + 2 * µ1**3
        if k == 4:
            return µ4 - 4 * µ1 * µ3 + 6 * µ1**2 * µ2 - 3 * µ1**4

        return None

    @property
    def mean(self) -> float:
        """
        Parametric mean
        """
        µ1 = self.non_central_moments(1)
        return self.loc + self.scale * µ1

    @property
    def variance(self) -> float:
        """
        Parametric variance
        """
        µ1 = self.non_central_moments(1)
        µ2 = self.non_central_moments(2)
        return self.scale**2 * (µ2 - µ1**2)

    @property
    def standard_deviation(self) -> float:
        """
        Parametric standard deviation
        """
        return numpy.sqrt(self.variance)

    @property
    def skewness(self) -> float:
        """
        Parametric skewness
        """
        central_µ3 = self.central_moments(3)
        µ1 = self.non_central_moments(1)
        µ2 = self.non_central_moments(2)
        std = numpy.sqrt(µ2 - µ1**2)
        return central_µ3 / std**3

    @property
    def kurtosis(self) -> float:
        """
        Parametric kurtosis
        """
        central_µ4 = self.central_moments(4)
        µ1 = self.non_central_moments(1)
        µ2 = self.non_central_moments(2)
        std = numpy.sqrt(µ2 - µ1**2)
        return central_µ4 / std**4

    @property
    def median(self) -> float:
        """
        Parametric median
        """
        return self.ppf(0.5)

    @property
    def mode(self) -> float:
        """
        Parametric mode
        """
        return self.loc + self.scale * (self.alpha / (self.alpha + 1)) ** (1 / self.alpha)

    @property
    def num_parameters(self) -> int:
        """
        Number of parameters of the distribution
        """
        return len(self.parameters)

    def parameter_restrictions(self) -> bool:
        """
        Check parameters restrictions
        """
        v1 = self.alpha > 0
        v2 = self.scale > 0
        return v1 and v2

    def get_parameters(self, continuous_measures) -> dict[str, float | int]:
        """
        Calculate proper parameters of the distribution from sample continuous_measures.
        The parameters are calculated by formula.

        Parameters
        ==========
        continuous_measures: MEASUREMESTS
            attributes: mean, std, variance, skewness, kurtosis, median, mode, min, max, size, num_bins, data

        Returns
        =======
        parameters: {"alpha": *, "loc": *, "scale": *}
        """

        def equations(initial_solution: tuple[float], continuous_measures) -> tuple[float]:
            alpha, loc, scale = initial_solution

            E = lambda k: scipy.special.gamma(1 - k / alpha)

            parametric_mean = E(1) * scale + loc
            parametric_variance = (scale**2) * (E(2) - E(1) ** 2)
            parametric_skewness = (E(3) - 3 * E(2) * E(1) + 2 * E(1) ** 3) / ((E(2) - E(1) ** 2)) ** 1.5
            # parametric_kurtosis = (E(4) - 4 * E(1) * E(3) + 6 * E(1) ** 2 * E(2) - 3 * E(1) ** 4) / ((E(2) - E(1) ** 2)) ** 2
            # parametric_median = loc + scale * (-numpy.log(0.5)) ** (-1 / alpha)
            # parametric_mode = loc + scale * (alpha / (alpha + 1)) ** (1 / alpha)

            ## System Equations
            eq1 = parametric_mean - continuous_measures.mean
            eq2 = parametric_variance - continuous_measures.variance
            eq3 = parametric_skewness - continuous_measures.skewness
            # eq3 = parametric_kurtosis - continuous_measures.kurtosis
            # eq2 = parametric_median - continuous_measures.median
            # eq3 = parametric_mode - continuous_measures.mode

            return (eq1, eq2, eq3)

        # x0 = (continuous_measures.mean, continuous_measures.mean, continuous_measures.mean)
        # bounds = ((0, -numpy.inf, 0), (numpy.inf, numpy.inf, numpy.inf))
        # args = [continuous_measures]
        # solution = scipy.optimize.least_squares(equations, x0=x0, bounds=bounds, args=args)
        # parameters = {"alpha": solution.x[0], "loc": solution.x[1], "scale": solution.x[2]}

        scipy_parameters = scipy.stats.invweibull.fit(continuous_measures.data_to_fit)
        parameters = {"alpha": scipy_parameters[0], "loc": scipy_parameters[1], "scale": scipy_parameters[2]}
        return parameters


if __name__ == "__main__":
    import sys

    sys.path.append("../")
    from continuous_measures import ContinuousMeasures

    ## Import function to get continuous_measures
    def get_data(path: str) -> list[float]:
        sample_distribution_file = open(path, "r")
        data = [float(x.replace(",", ".")) for x in sample_distribution_file.read().splitlines()]
        sample_distribution_file.close()
        return data

    ## Distribution class
    path = "../continuous_distributions_sample/sample_frechet.txt"
    data = get_data(path)
    continuous_measures = ContinuousMeasures(data)
    distribution = Frechet(continuous_measures=continuous_measures)

    print(f"{distribution.name} distribution")
    print(f"Parameters: {distribution.parameters}")
    print(f"CDF: {distribution.cdf(continuous_measures.mean)} {distribution.cdf(numpy.array([continuous_measures.mean, continuous_measures.mean]))}")
    print(f"PDF: {distribution.pdf(continuous_measures.mean)} {distribution.pdf(numpy.array([continuous_measures.mean, continuous_measures.mean]))}")
    print(f"PPF: {distribution.ppf(0.5)} {distribution.ppf(numpy.array([0.5, 0.5]))} - V: {distribution.cdf(distribution.ppf(0.5))}")
    print(f"SAMPLE: {distribution.sample(5)}")
    print(f"\nSTATS")
    print(f"mean: {distribution.mean} - {continuous_measures.mean}")
    print(f"variance: {distribution.variance} - {continuous_measures.variance}")
    print(f"skewness: {distribution.skewness} - {continuous_measures.skewness}")
    print(f"kurtosis: {distribution.kurtosis} - {continuous_measures.kurtosis}")
    print(f"median: {distribution.median} - {continuous_measures.median}")
    print(f"mode: {distribution.mode} - {continuous_measures.mode}")
